### PR TITLE
[mdns] add native mDNS support in OpenThread

### PR DESCRIPTION
**Pull request by abtink**
This commit introduces native mDNS support within the OpenThread stack, implementing RFC 6762 compliant registration of hosts, services, and keys. It supports the following functionalities:

- Sending probes to claim names.
- Sending announcements on initial registration and changes.
- Sending "goodbye" announcements when unregistered or upon record removal.
- Negative responses (NSEC).
- Support for service sub-types and their addition/removal.
- Support for `_services._dns-sd._udp` queries (all service types).
- Responding to queries (including "QU" questions requesting unicast response).
- Delay mechanism when responding to multi-question query messages, for unique or shared records.
- Providing extra records in the Additional Data section if not already in the Answer section (e.g., on a PTR query, include SRV and host AAAA addresses).
- Implementing Known-Answer Suppression.
- Supporting multi-packet queries with known answers in follow-up messages.
- Rate-limiting record multicasts (once per second).
- Rate-limiting probe responses (once per 250ms).
- Detecting conflicts after probes.
- Limiting the size of emitted responses or probes, breaking into multiple messages if necessary.
- Detecting self originating messages (sent by mDNS module).
- Support for service browser.
- Support for service resolvers (SRV and TXT records) and IPv4/IPv6 address resolvers for hostnames.
- Introduces smart cache management:
  - Passively caches service records for active browsers.
  - Passively caches address records for active service resolvers.
- Enables multiple simultaneous browsers/resolvers for the same service/host.

This commit introduces public `otMdns` OpenThread APIs and related CLI
commands for the mDNS module.

For platform abstraction, `otPlatMdns` APIs are defined in `mdns_socket.h` (e.g., to send or receive mDNS messages):

- An implementation of the platform APIs is provided for posix.
- Also under the simulation platform, a simplified implementation of the `otPlatMdns` APIs is provided (intended for testing).
This commit also adds a detailed `test_mdns` unit test, validating various functionalities and covering potential edge cases.

---------------------------------

~This PR is the first phase of adding native mDNS support in OT core adding support for registering entries. The next phase (in follow-up PR) will add support for service browsers and service/address resolvers.~

The PR now contains the changes from next phase (adding support for browser and service/address resolvers). The new changes will be available in this https://github.com/openthread/openthread/pull/9865 as a separate commit for some time (so to allow easier review of the new code).

Some functionalities are left for future enhancements/PRs:
- Legacy unicast query (source port not matching mDNS port)
- Direct unicast queries to mDNS port

### CODE REVIEW

**Comment by wgtdkp** [comment]
@abtink The OpenThread mDNS will ever be enabled on a Border Router, right? It's common that there is already platform mDNS server which is leveraged by multiple software components on a Border Router devices, so it will be hard for Thread vendors to transition to OpenThread mDNS. Is it expected that BRs will have two mDNS servers running simutaneously? Is this technically doable?

File: include/openthread/platform/mdns_socket.h [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+ *
+ * All received messages MUST be reported back using `otPlatMdnsHandleReceive()` callback.
+ *
+ * @param[in] aInstance  The OpernThread instance.
+ * @param[in] aEnable    Indicate whether to enable or disable.
+ *
+ */
+void otPlatMdnsSetEnabled(otInstance *aInstance, bool aEnable);
+
+/**
```
Comment (wgtdkp) [inline include/openthread/platform/mdns_socket.h:90]: This API doesn't specify on which interface to send the multicast packet

File: src/core/net/mdns.cpp [hunk 2]
```diff
@@ -0,0 +1,3881 @@
...
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "mdns.hpp"
+
+#if OPENTHREAD_CONFIG_MULTICAST_DNS_ENABLE
```
Comment (wgtdkp) [inline src/core/net/mdns.cpp:31]: Can this be enabled when Border Router features are not enabled? If yes, but on which interface to advertise the mDNS services?

**Comment by abtink** [comment]
> @abtink The OpenThread mDNS will ever be enabled on a Border Router, right? 

Yes. Thread BR is the primary use-case. 

> It's common that there is already platform mDNS server which is leveraged by multiple software components on a Border Router devices, so it will be hard for Thread vendors to transition to OpenThread mDNS. Is it expected that BRs will have two mDNS servers running simutaneously? Is this technically doable?

Yes. It is possible to have more than one mDNS responder and/or querier implementation coexist on the same device. RFC [6762 - section 15](https://datatracker.ietf.org/doc/html/rfc6762#section-15) covers this. 

---

#### Background on this PR:

OpenThread currently delegates mDNS functionality (including related adv-proxy and discovery-proxy functions) to the platform layer. We will continue to support this model.

Our goal is to provide native support for these functions within the OpenThread core stack, offering an alternative option for platforms and products.


#### Regarding mDNS function:

The current model delegates mDNS functionality to the platform layer. For the `posix` platform (in `ot-br-posix`), we define the `otbr::Mdns::Publisher` class as an abstraction layer. We also offer example integrations with mDNSResponder and Avahi, but platforms can choose other mDNS libraries.

Advantages of native support:

- **Platform-agnostic**: Native mDNS support within OpenThread would be usable across different platforms and operating systems.
- **Improved reliability**: A well-tested, Thread-specific mDNS implementation within OpenThread core could reduce issues related to BR service discovery that stem from platform-specific mDNS libraries.
- **Faster feature development and adoption**: Adding new mDNS features or behaviors (e.g., [TSR](https://datatracker.ietf.org/doc/draft-tllq-tsr/) record for conflict detection) would be streamlined with native support.

File: include/openthread/platform/mdns_socket.h (hunk 1 above)
Comment (abtink) [reply include/openthread/platform/mdns_socket.h:90]: This API requires the platform to multicast the message over all network interfaces within the mDNS multicast domain. The platform is responsible to manage the set of relevant interfaces.

My idea is that if a future use case arises where we need to multicast on a specific interface, we introduce a new separate API for that purpose. This would make this behavior explicit and clear.

File: src/core/net/mdns.cpp (hunk 2 above)
Comment (abtink) [reply src/core/net/mdns.cpp:31]: Technically yes. The platform is responsible to manage the set of relevant interfaces for mDNS.

If we realize we need OT core to be aware of set of interfaces, we can update this. Do you see any use-case for this?

File: include/openthread/platform/mdns_socket.h (hunk 1 above)
Comment (wgtdkp) [reply include/openthread/platform/mdns_socket.h:90]: Border routing currently supports only a single link, if a SRP service is advertised on all mdns-capable links, the service will be discoverable on a non-AIL link but **not reachable**. I think this is worse than making SRP service discoverable on only AIL link. So for SRP, it's better to specify the desired the advertising interface. Thoughts?

Comment (abtink) [reply include/openthread/platform/mdns_socket.h:90]: I agree. Maybe to phrase it better we should say "this API should multicast on all network interfaces where device is acting as a Border Router".  Today that is limited to one interface. I will add an parameter to specify the interface index in the next push to this API.

Comment (abtink) [reply include/openthread/platform/mdns_socket.h:88]: @wgtdkp the new pushed commit adds the interface index to this platform API and other APIs.

Comment (wgtdkp) [reply include/openthread/platform/mdns_socket.h:91]: Thanks abtink! resolving this comment.

**Comment by abtink** [comment]
> Besides the unittests what other testing has been used to validate the mdns implementation?

@alexeRadu , the unit test validates operation under normal conditions and various edge cases. 

Additionally, this PR provides an `otPlatMdns` platform API implementation under the simulation platform (using POSIX socket). This is enabled by `OPENTHREAD_SIMULATION_MDNS_SCOKET_IMPLEMENT_POSIX`, and allows OpenThread stack executables (built with the simulation platform) to run directly on standard machines (laptop/desktop), acting as an mDNS entity on the network. We've used this feature for manual testing and validation of OpenThread's mDNS implementation against others (primarily mDNSResponder on macOS via `dns-sd` commands). Wireshark was also used to capture and analyze the emitted mDNS messages.

Our long-term goal is add similar automated tests (as part of OpenThread CI), either leveraging `OPENTHREAD_SIMULATION_MDNS_SCOKET_IMPLEMENT_POSIX` or exploring new mechanisms. 

Further contributions to the implementation's testing are always welcome. :)

File: src/core/net/mdns.hpp [hunk 3]
```diff
@@ -0,0 +1,1118 @@
...
+    static const char kLocalDomain[];         // "local."
+    static const char kUdpServiceLabel[];     // "_udp"
+    static const char kTcpServiceLabel[];     // "_tcp"
+    static const char kSubServiceLabel[];     // "_sub"
+    static const char kServciesDnssdLabels[]; // "_services._dns-sd._udp"
+
+    bool                     mIsEnabled;
+    bool                     mIsQuestionUnicastAllowed;
+    uint16_t                 mMaxMessageSize;
+    uint32_t                 mInfraIfIndex;
```
Comment (radua-nxp) [inline src/core/net/mdns.hpp:1093]: I'm guessing that mInfraIfIndex is an interface number. But it seems that is always used by the otPlatMdnsXyz() functions. If so why doesn't the platform implementation keeps track of the interface number?

Comment (abtink) [reply src/core/net/mdns.hpp:1093]: We want it to be similar to existing `otPlatInfraIf` APIs (where `aInfraIndex` is used in APIs). Including the infra-if index in the APIs will ensure we can extend/enhance the behavior in future.

Also please check related discussion in this comment https://github.com/openthread/openthread/pull/9797#discussion_r1466354688

File: src/core/net/mdns.hpp [hunk 4]
```diff
@@ -0,0 +1,1118 @@
...
+        // `AppendState` methods: Used to track whether the record
+        // is appended in a message, or needs to be appended in
+        // Additional Data section.
+
+        void MarkAsNotAppended(void) { mAppendState = kNotAppended; }
+        void MarkAsAppended(TxMessage &aTxMessage, Section aSection);
+        void MarkToAppendInAdditionalData(void);
+        bool IsAppended(void) const;
+        bool CanAppend(void) const;
+        bool ShouldAppnedInAdditionalDataSection(void) const { return (mAppendState == kToAppendInAdditionalData); }
```
Comment (Cristib05) [inline src/core/net/mdns.hpp:536]: small typo here

File: src/core/net/mdns.hpp (hunk 3 above)
Comment (radua-nxp) [reply src/core/net/mdns.hpp:1093]: ok, got it; thanks

File: src/core/net/mdns.cpp [hunk 5]
```diff
@@ -0,0 +1,3887 @@
...
+    case kRemoving:
+        break;
+    }
+}
+
+void Core::Entry::StartProbing(void)
+{
+    SetState(kProbing);
+    mProbeCount = 0;
+    SetFireTime(TimerMilli::GetNow() + kInitalProbeDelay);
```
Comment (radua-nxp) [inline src/core/net/mdns.cpp:908]: RFC 6762 8.1 Probing - When ready to send its Multicast DNS probe packet(s) the host should
   first wait for a short random delay time, uniformly distributed in
   the range 0-250 ms.  This random delay is to guard against the case
   where several devices are powered on simultaneously, or several
   devices are connected to an Ethernet hub, which is then powered on,
   or some other external event happens that might cause a group of
   hosts to all send synchronized probes.

File: src/core/net/mdns.cpp [hunk 6]
```diff
@@ -0,0 +1,3887 @@
...
+    EntryType *entry;
+
+    VerifyOrExit(mIsEnabled, error = kErrorInvalidState);
+
+    entry = GetList<EntryType>().FindMatching(aItemInfo);
+
+    if (entry == nullptr)
+    {
+        entry = EntryType::AllocateAndInit(GetInstance(), aItemInfo);
+        OT_ASSERT(entry != nullptr);
```
Comment (radua-nxp) [inline src/core/net/mdns.cpp:131]: Isn't it going to be a race-condition between the Probing (started in AllocateAndInit() -> Entry::Init()) and GetList<EntryType().Push(*entry)? Basically what if the probing timer expires before we push to the list?

Comment (abtink) [reply src/core/net/mdns.cpp:131]: All OpenThread API calls and platform callbacks (including timer/alarm) should run from single OS context (OT stack is not re-entrant). 

> Basically what if the probing timer expires before we push to the list?

Even if timer expires, platform must ensure to signal alarm callback `otPlatAlarmMilliFired()` after current `Tasklet` and/or call to OT stack has returned.

File: src/core/net/mdns.hpp (hunk 4 above)
Comment (abtink) [reply src/core/net/mdns.hpp]: Fixed in new push. Thanks.

File: src/cli/cli_mdns.cpp [hunk 7]
```diff
@@ -0,0 +1,776 @@
...
+void Mdns::HandleRegisterationDone(otMdnsRequestId aRequestId, otError aError)
+{
+    if (mWaitingForCallback && (aRequestId == mRequestId))
+    {
+        mWaitingForCallback = false;
+        Interpreter::GetInterpreter().OutputResult(aError);
+    }
+    else
+    {
+        OutputLine("mDNS registration for request id %u outcome: %s", aRequestId, otThreadErrorToString(aError));
```
[the remark this answers is not in the document]
Comment (abtink) [reply src/cli/cli_mdns.cpp]: Yes. Thanks. Updated in new pushed.

**Comment by abtink** [comment]
New pushed commit (will be squashed later):
```
(to squash later) enhance simulation/mdns_socket.c
- Create IPv6 socket in addition to IPv4
- Send and receive mDNS messages on both IPv6 and IPv4
- Simplify logging and handling errors
- Implement `otPlatMdnsSendUnicast()`
```

**Comment by abtink** [comment]
The changes from the PR below are now included in the commit in this PR:
- https://github.com/openthread/openthread/pull/9935

**Comment by marius-preda** [comment]
HI @abtink, when do you think this PR will be merged? The mDNS implementation seems quite complete. We have tested it and it works ok in our setup. I see you are adding some other PRs connected to this, do you have more in the pipeline?

**Comment by abtink** [comment]
> HI @abtink, when do you think this PR will be merged? The mDNS implementation seems quite complete. We have tested it and it works ok in our setup. I see you are adding some other PRs connected to this, do you have more in the pipeline?

Good to hear it's been validated in your setup. It would be helpful to mark the PR approved (and reviewed). Thanks.

There are some other PR building on to of this (some are squashed into this):
- https://github.com/openthread/openthread/pull/9935
- https://github.com/openthread/openthread/pull/9944
- https://github.com/openthread/openthread/pull/9954

Some other related ones in pipeline:
- Updating disc proxy to use the new native mDNS
- Adding APIs to iterate over active browsers, resolvers (and/or cache) - intended for testing/debugging.
- Later on we may add support for "Legacy unicast query (source port not matching mDNS port)" and 
"Direct unicast queries to mDNS port" (though these are lower priority).

File: src/core/net/mdns.cpp [hunk 8]
```diff
@@ -0,0 +1,6012 @@
...
+    , mNextProbeTxTime(TimerMilli::GetNow() - 1)
+    , mEntryTimer(aInstance)
+    , mEntryTask(aInstance)
+    , mTxMessageHistory(aInstance)
+    , mConflictCallback(nullptr)
+    , mNextQueryTxTime(TimerMilli::GetNow() - 1)
+    , mCacheTimer(aInstance)
+    , mCacheTask(aInstance)
+
+{
```
Comment (jwhui) [inline src/core/net/mdns.cpp:87]:
```suggestion
    , mCacheTask(aInstance)
{
```

File: include/openthread/mdns.h [hunk 9]
```diff
@@ -0,0 +1,747 @@
...
+ *
+ */
+void otMdnsSetConflictCallback(otInstance *aInstance, otMdnsConflictCallback aCallback);
+
+/**
+ * Registers or updates a host on mDNS.
+ *
+ * The fields in @p aHost follow these rules:
+ *
+ * - The `mHostName` field specifies the host name to register (e.g., "myhost"). MUST not contain the domain name.
```
[the remark this answers is not in the document]
Comment (abtink) [reply include/openthread/mdns.h]: Changed other instances of `MUST not` -> `MUST NOT`

File: include/openthread/platform/mdns_socket.h [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+ *
+ */
+otError otPlatMdnsSetListeningEnabled(otInstance *aInstance, bool aEnable, uint32_t aInfraIfIndex);
+
+/**
+ * Sends an mDNS message as multicast.
+ *
+ * The platform MUST multicast the prepared mDNS message in @p aMessage as a UDP message using the mDNS well-known port
+ * number 5353 for both source and destination ports. The message MUST be sent to the mDNS link-local multicast
+ * address `224.0.0.251` and/or its IPv6 equivalent `ff02::fb`.
```
[the remark this answers is not in the document]
Comment (abtink) [reply include/openthread/platform/mdns_socket.h:96]: Change it to "and"

**Review by Cristib05 — APPROVED** [review]
In terms of functionality, LGTM. Also, tested the implementation on a hardware platform, works as expected.

**Review by radua-nxp — APPROVED** [review]

**Comment by Vyrastas** [comment]
@jrhodie can you review the documentation portion of this please.

File: include/openthread/platform/mdns_socket.h [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+ *
+ * @p aMessage contains the mDNS message starting with DNS header at offset zero. It does not include IP or UDP headers.
+ * This function passes the the ownership of @p aMessage to the platform layer and platform implementation MUST free
```
Comment (jrhodie) [inline include/openthread/platform/mdns_socket.h:99]: double "the  the"

File: include/openthread/platform/mdns_socket.h [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+void otPlatMdnsSendMulticast(otInstance *aInstance, otMessage *aMessage, uint32_t aInfraIfIndex);
+
+/**
+ * Sends an mDNS message as unicast.
+ *
+ * The platform MUST send the prepared mDNS message in @p aMessage as a UDP message using source UDP port 5353 to
+ * the destination address and port number specified by @p aAddress.
+ *
+ * @p aMessage contains the DNS message starting with the DNS header at offset zero. It does not include IP or UDP
+ * headers. This function passes the the ownership of @p aMessage to the platform layer and platform implementation
```
Comment (jrhodie) [inline include/openthread/platform/mdns_socket.h:121]: there is a double "the  the"

File: include/openthread/platform/mdns_socket.h [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+ * @param[in] aAddress    The destination address info.
+ *
+ */
+void otPlatMdnsSendUnicast(otInstance *aInstance, otMessage *aMessage, const otPlatMdnsAddressInfo *aAddress);
+
+/**
+ * Callback to notify OpenThread mDNS module of a received message on UDP port 5353.
+ *
+ * @p aMessage MUST contain DNS message starting with the DNS header at offset zero. This function passes the the
+ * ownership of @p aMessage from the platform layer to the OpenThread stack. The OpenThread stack will free the
```
Comment (jrhodie) [inline include/openthread/platform/mdns_socket.h:144]: there is a double "the  the"

File: include/openthread/platform/mdns_socket.h [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+} otPlatMdnsAddressInfo;
+
+/**
+ * Enables or disables listening for mDNS messages sent to mDNS port 5353.
+ *
+ * When listening is enabled, platform MUST listen for multicast messages sent to UDP destination port 5353 at the
+ * mDNS link-local multicast address `224.0.0.251` and its IPv6 equivalent `ff02::fb`.
+ *
+ * Platform SHOULD also listen for any unicast messages sent to UDP destination port 5353. If this is not possible,
+ * then OpenThread mDNS module can be configured so to not use any "QU" questions requesting unicast response.
```
Comment (jrhodie) [inline include/openthread/platform/mdns_socket.h:77]: suggest removing the word "so" above.

**Review by jrhodie — COMMENTED** [review]
just a few very minor things....

File: include/openthread/platform/mdns_socket.h (hunk 14 above)
Comment (abtink) [reply include/openthread/platform/mdns_socket.h]: Thanks. Updated all in new push.

**Review by jrhodie — APPROVED** [review]

File: include/openthread/mdns.h [hunk 15]
```diff
@@ -0,0 +1,747 @@
...
+/**
+ * Represents an mDNS key record.
+ *
+ * See `otMdnsRegisterKey()`, `otMdnsUnregisterKey()` for more details about fields in each case.
+ *
+ */
+typedef otPlatDnssdKey otMdnsKey;
+
+/**
+ * Enables or disables mDNS module.
```
Comment (Vyrastas) [inline include/openthread/mdns.h:137]:
```suggestion
 * Enables or disables the mDNS module.
```

File: include/openthread/mdns.h [hunk 16]
```diff
@@ -0,0 +1,747 @@
...
+ *
+ * mDNS module should be enabled before registration any host, service, or key entries. Disabling mDNS will immediately
```
Comment (Vyrastas) [inline include/openthread/mdns.h:139]:
```suggestion
 * The mDNS module should be enabled before registration any host, service, or key entries. Disabling mDNS will immediately
```

File: include/openthread/mdns.h [hunk 17]
```diff
@@ -0,0 +1,747 @@
...
+ * where platform socket cannot accept unicast response received on mDNS port (due to it being already bound).
+ *
+ * @param[in] aInstance     The OpenThread instance.
+ * @param[in] aAllow        Indicates whether or not to allow "QU" questions.
+ *
+ */
+void otMdnsSetQuestionUnicastAllowed(otInstance *aInstance, bool aAllow);
+
+/**
+ * Indicates whether mDNS module is allowed to send "QU" questions requesting unicast response.
```
Comment (Vyrastas) [inline include/openthread/mdns.h:180]:
```suggestion
 * Indicates whether the mDNS module is allowed to send "QU" questions requesting unicast response.
```

**Review by Vyrastas — CHANGES_REQUESTED** [review]

File: src/cli/cli_mdns.cpp [hunk 18]
```diff
@@ -0,0 +1,776 @@
...
+        OutputLine("Key %s (host)", aKey.mName);
+    }
+
+    OutputFormat(kIndentSize, "key-data: ");
+    OutputBytesLine(aKey.mKeyData, aKey.mKeyDataLength);
+
+    OutputLine(kIndentSize, "ttl: %lu", ToUlong(aKey.mTtl));
+}
+
+template <> otError Mdns::Process<Cmd("register")>(Arg aArgs[])
```
[the remark this answers is not in the document]
Comment (abtink) [reply src/cli/cli_mdns.cpp:151]: Not yet. At a later point and in a seperate PR.

The native mDNS implementation is intended for Thread and OT internal uses-cases (e.g., Advertising Proxy, Discovery Proxy, ect). We want to discourage its use for other use-cases. The public API or CLI for mDNS are mainly intended for testing, so no rush to document them.

File: include/openthread/platform/mdns_socket.h [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * @brief
+ *   This file includes the platform abstraction for mDNS socket.
```
[the remark this answers is not in the document]
Comment (abtink) [reply include/openthread/platform/mdns_socket.h:32]: Updated in new push.

File: src/core/config/mdns.h [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
```
[the remark this answers is not in the document]
Comment (abtink) [reply src/core/config/mdns.h:30]: Added in new push. Thanks.

File: include/openthread/mdns.h (hunk 17 above)
Comment (abtink) [reply include/openthread/mdns.h:180]: Updated all in new push. Thanks.

File: src/core/net/mdns.cpp [hunk 21]
```diff
@@ -0,0 +1,6011 @@
...
+    EntryType *entry;
+
+    VerifyOrExit(mIsEnabled, error = kErrorInvalidState);
+
+    entry = GetEntryList<EntryType>().FindMatching(aItemInfo);
+
+    if (entry == nullptr)
+    {
+        entry = EntryType::AllocateAndInit(GetInstance(), aItemInfo);
+        OT_ASSERT(entry != nullptr);
```
Comment (konradderda) [inline src/core/net/mdns.cpp:142]: 1. I cannot find validation of provided names before memory allocation, e.g. we can provide names like `...local.`
2. It feels that asserting if `AllocateAndInit()` is unable to allocate a memory for entry is a little too harsh. Maybe we should just print an error?

File: src/core/net/mdns.cpp [hunk 22]
```diff
@@ -0,0 +1,6011 @@
...
+    }
+
+    for (const Question &question : mQuestions)
+    {
+        AnswerQuestion(question, answerTime);
+    }
+
+    if (needUnicastResponse)
+    {
+        SendUnicastResponse(mSenderAddress);
```
Comment (konradderda) [inline src/core/net/mdns.cpp:3294]: Unicast response should have the same Transaction ID as the Query

**Review by konradderda — COMMENTED** [review]
I have left few remarks as first iteration

File: src/core/net/mdns.cpp [hunk 23]
```diff
@@ -0,0 +1,6011 @@
...
+{
+    return (aTtl == kUnspecifiedTtl) ? aDefaultTtl : aTtl;
+}
+
+bool Core::NameMatch(const Heap::String &aHeapString, const char *aName)
+{
+    // Compares a DNS name given as a `Heap::String` with a
+    // `aName` C string.
+
+    return !aHeapString.IsNull() && StringMatch(aHeapString.AsCString(), aName, kStringCaseInsensitiveMatch);
```
[the remark this answers is not in the document]
Comment (abtink) [reply src/core/net/mdns.cpp:333]: Yes. Leaving as future TODO sounds good.

File: src/core/net/mdns.cpp (hunk 21 above)
Comment (abtink) [reply src/core/net/mdns.cpp:142]: Thanks for the review (and noticing the typos :)). All fixed in new push.

Both these are intentional design choices for the current implementation of native mDNS in OT to prioritize code simplicity and reduce overhead.
- Regarding checking the names, the native OT mDNS is primarily intended for use by the OT core stack (advertising proxy, discovery proxies, TREL, etc.) where names are expected to be pre-validated.
- Regarding heap allocation failure: We debated this internally and decided on the current model. Handling heap allocation errors in all situations can become complex, requiring significant code to roll back state and release other resources.  Practically, heap allocation failures are often due to memory leaks, where an assert/reset may be more appropriate than letting the device/BR continue to run with degraded functionality. We can revisit this and possibly add heap allocation error-handling mechanisms if the need becomes pressing.

File: src/core/net/mdns.cpp (hunk 22 above)
Comment (abtink) [reply src/core/net/mdns.cpp:3294]: Thanks @konradderda.

I understand the ID must be mirrored in responses to Direct Unicast Queries (RFC 6762 - section 5.5) and Legacy Unicast queries (section 6.7). However, the current implementation does not support handling unicast queries (left for future enhancement).

This is a response to an mDNS "QU" query question. Not sure if we need to mirror ID for this? Can you please maybe point me to the RFC text for this? Thanks.

Comment (konradderda) [reply src/core/net/mdns.cpp:3294]: Honestly, when I was writing the comment I thought it is true for all unicast responses. I recalled my fight with ot-commissioner which expects unicast responses sent to an ephemeral port while using `borderagent discover` command.

This cannot be changed on ot-commissioner's side as it must not use 5353 source port (section 5.1). It seems that support for Legacy Unicast Queries is needed sooner than later (at least for external commissioning).

File: src/core/net/mdns.cpp (hunk 21 above)
Comment (konradderda) [reply src/core/net/mdns.cpp:142]: Ad 1. I understand the reasoning. Maybe some kind of @note should be added to the otMdnsRegister* APIs to inform an user about a need of prevalidation?

Ad 2. The problem with this particular assert is that for constrained devices it can be easly triggered by registering a lot of services. It feels like we could just return an Error that could be passed to the other party (e.g. SRP client) instead of resetting the device which could cause a re-registration of all services again and potentially lead to a loop.

**Review by konradderda — APPROVED** [review]
The code looks good, thanks

**Review by Vyrastas — APPROVED** [review]

**Review by jwhui — APPROVED** [review]
